jitbuilder/lljb/src/IRVisitor.cpp

```diff
@@ -0,0 +1,559 @@
+   llvm::Function * callee = I.getCalledFunction();
+   const char * calleeName = callee->getName().data();
+   std::size_t numParams = I.arg_size();
+   std::vector<TR::IlValue *> params;
+   std::vector<TR::IlType *> paramTypes;
+   for (int i = 0; i < numParams; i++)
+      {
+      TR::IlValue * parameter = getIlValue(I.getArgOperand(i));
+      params.push_back(parameter);
+      paramTypes.push_back(_methodBuilder->getIlType(_td,I.getArgOperand(i)->getType()));
+      }
+   _methodBuilder->defineFunction(callee,numParams, paramTypes.data());
+   result = _builder->Call(calleeName, numParams, params.data());
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitCastInst(llvm::CastInst &I)
+   {
+   TR::IlValue * srcVal = getIlValue(I.getOperand(0));
+   TR::IlType * toIlType = _methodBuilder->getIlType(_td,I.getDestTy());
+   TR::IlValue * result = _builder->ConvertTo(toIlType, srcVal);
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitZExtInst(llvm::ZExtInst &I)
+   {
+   TR::IlValue * srcVal = getIlValue(I.getOperand(0));
+   TR::IlType * toIlType = _methodBuilder->getIlType(_td,I.getDestTy());
+   TR::IlValue * result = _builder->UnsignedConvertTo(toIlType, srcVal);
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitGetElementPtrInst(llvm::GetElementPtrInst &I)
+   {
+   TR::IlValue * ilValue = nullptr;
+   if (I.getSourceElementType()->isStructTy())
+      {
+      llvm::ConstantInt * indextConstantInt = llvm::dyn_cast<llvm::ConstantInt>(I.getOperand(2));
+      unsigned elementIndex = indextConstantInt->getZExtValue();
+      ilValue = _builder->StructFieldInstanceAddress(I.getSourceElementType()->getStructName().data(),
+                                                               _methodBuilder->getMemberNameFromIndex(elementIndex),
+                                                               getIlValue(I.getOperand(0)));
+      }
+   else if (I.getSourceElementType()->isArrayTy())
+      {
+      if (I.getSourceElementType()->getArrayElementType()->isArrayTy())
+         {
+         llvm::ConstantInt * indextConstantInt = llvm::dyn_cast<llvm::ConstantInt>(I.getOperand(2));
+         int64_t elementIndex = indextConstantInt->getSExtValue();
+         elementIndex *= I.getSourceElementType()->getArrayNumElements();
+         ilValue =
+               _builder->IndexAt(
+                  _td->PointerTo(
+                     _methodBuilder->getIlType(_td,I.getSourceElementType()->getArrayElementType())),
+                  getIlValue(I.getOperand(0)),
+                  _builder->ConstInt32(elementIndex));
+         }
+      else
+         {
+         TR::IlValue * elementIndex = getIlValue(I.getOperand(2));
+         ilValue =
+               _builder->IndexAt(
+                  _td->PointerTo(
+                     _methodBuilder->getIlType(_td,I.getSourceElementType()->getArrayElementType())),
+                  getIlValue(I.getOperand(0)),
```

Comment:
another `IndexAt` with index 0

Reply:
This is not `IndexAt` at index 0 similar to the Load/Store instruction visitors. `getelementptr` instruction is used for address computations involving aggregate data types. In this particular line where we deal with single dimension arrays, the second argument provided to `IndexAt` is `getIlValue(I.getOperand(0))` which gets the `IlValue` corresponding to the array allocated in an earlier instruction. `getOperand(0)` on `getelementptr` instruction gets the `llvm::Value` mapped to the `IlValue` we want here.